### PR TITLE
Correcting asynchronous testing

### PR DESCRIPTION
**Pull request by mhoyer**
Investigating the test suite of TodoMVC I found missing asynchronicity of the tests (missing callbacks) and the actual chaining of promises across ```page.js```, ```testOperations.js```, and inside the tests itself. Hence I decided to fix this. All in all I applied these improvements/fixes:

- made all tests (```it()```s) to actually be asynchronous tests by adding and handling a ```done``` callback
- promisified all functions in ```page.js```, ```testOperations.js```, and ```tests.js```:
  - actually ```return```ing the promises
  - chaining the promises with ```.then()```s
  - flattening out some ```.then()``` nestings
- extracted new ```page.waitForVisibleElement()```
- changed ```page.ensureAppIsVisible()``` to wait for ```new-todo``` input box instead of ```todoapp```. Some SPAs are dynamically loading the inner content where it might happen that first tests fail due to timing issues.
- simplified test setup for speed-mode and non-speed-mode
- demystified the ```idSelectors``` variable by inverting its meaning and kicking the general ```!```-negation

### CODE REVIEW

File: tests/page.js [hunk 1]
```diff
@@ -175,9 +175,24 @@ module.exports = function Page(browser) {
 		});
 	};
 
+	this.waitForVisibleElement = function (getElementFn, timeout) {
+		var foundVisibleElement;
+		timeout = timeout || 500;
```
Comment (mhoyer) [inline tests/page.js]: The magic 500 is some kind of a hack at this place. I experienced slow tests when running the clear-complete-button tests against ```backbone``` framework. With Backbone approach the items gets hidden with the filter buttons whereas my Aurelia solution rebuilds the whole items list. Thus, for the backbone variant the ```waitForVisibleElement()``` helper will wait for hidden items to get visible in some cases.
:exclamation: Needs further investigation.

**Comment by ColinEberhardt** [comment]
Thanks @mhoyer - there are some really good improvements here, especially those in `page` and `testOperations`. However there are a number which are not strictly necessary.

The WebDriverJS API has a [promise manager](https://code.google.com/p/selenium/wiki/WebDriverJs#Control_Flows) which removes the need to explicitly chain promises.

Did you actually experience any problems with the runtime functionality of these tests? Or was your aim just to refactor / optimise?

**Comment by mhoyer** [comment]
> However there are a number which are not strictly necessary.
>
> The WebDriverJS API has a promise manager which removes the need to explicitly chain promises.

Interesting. Didn't know this fact. But IMHO having those explicit ```then()```-chains seems less magic to me. Maybe others also don't know about this queueing approach in ```selenium-webdriver```. But I am also fine with removing the unnecessary ```then()```s.

:question: What do you (or others) think? @samccone? 

> Did you actually experience any problems with the runtime functionality of these tests? Or was your aim just to refactor / optimise?

Indeed, I had issues with my Aurelia implementation. E. g. the *"Clear completed button - should remove completed items when clicked"* where Aurelia seems to render the visibility of the clear completed button slightly later as the test runner executes the assertions.

**Comment by mhoyer** [comment]
I added one commit that reverts the unnecessary ```.then()```-chaining in ```test.js```. Alternatively we could introduce the [fat arrow syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) instead to reduce this ```function ()``` noise and to keep the explicit promise handling.

    createStandardItems()
      .then(function () { return testOps.assertItemCount(3); })
      .then(function () { return testOps.assertItemText(0, TODO_ITEM_ONE); })
      .then(function () { return testOps.assertItemText(1, TODO_ITEM_TWO); })
      .then(function () { return testOps.assertItemText(2, TODO_ITEM_THREE); })

vs.

    createStandardItems()
      .then(() => testOps.assertItemCount(3))
      .then(() => testOps.assertItemText(0, TODO_ITEM_ONE))
      .then(() => testOps.assertItemText(1, TODO_ITEM_TWO))
      .then(() => testOps.assertItemText(2, TODO_ITEM_THREE))

**Comment by ColinEberhardt** [comment]
I'm happy to drop the promise 'magic' in favour of a more obvious implementation. The way that webdriver forces you into a world where everything is asynchronous is something I really dislike.

Before merging I would like to run a full set of tests - this will have to sit until Monday when I am back in the office.

Thanks a lot for these updates - really appreciate this.

**Comment by ColinEberhardt** [comment]
OK, so I've had a chance to do some more comprehensive testing, running these updated tests against a range of frameworks and it all looks good to me.

I think I prefer the updated implementation which makes use of the built-in promise manager. 

Basically this gets the :+1: from me.

**Comment by ColinEberhardt** [comment]
Merged via #1573

**Comment by mhoyer** [comment]
:sunglasses:

**Pull request closed without merging**